ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/DatabaseSelfServiceError.scala

```diff
@@ -0,0 +1,109 @@
+  *
+  * TODO error codes: Move and handle the error specialization per storage backend
+  *                   in [[com.daml.platform.store.backend.StorageBackend]].
+  */
+object DatabaseSelfServiceError {
+  @tailrec
+  def apply(
+      exception: Throwable
+  )(implicit contextualizedErrorLogger: ContextualizedErrorLogger): Throwable = exception match {
+    // This frequently occurs when running with H2, because H2 does not properly implement the serializable
+    // isolation level. This causes unexpected constraint violation exceptions when running with H2 in the presence
+    // of contention. For now, we retry on these exceptions.
+    // See https://github.com/h2database/h2database/issues/2167
+    case ex: JdbcSQLIntegrityConstraintViolationException => retryable(ex)
+    case ex: SQLRecoverableException => retryable(ex)
+    case ex: SQLTransientException => retryable(ex)
+    case ex: SQLNonTransientException => nonRetryable(ex)
+    case ex: PSQLException => if (isRetryablePsqlException(ex)) retryable(ex) else nonRetryable(ex)
+    // Oracle uses java.sql.SqlException and java.sql.BatchUpdateException
+    case ex: BatchUpdateException if ex.getCause != null => DatabaseSelfServiceError(ex.getCause)
+    case ex: SQLException => if (isRetryableOracleException(ex)) retryable(ex) else nonRetryable(ex)
+    // Don't handle other exceptions that can be thrown from non-client interactions (e.g. index initialization)
+    case ex => ex
+  }
+
+  private def retryable(ex: SQLException)(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): StatusRuntimeException =
+    IndexErrors.DatabaseErrors.SqlTransientError.Reject(ex).asGrpcError
+
+  private def nonRetryable(ex: SQLException)(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): StatusRuntimeException =
+    IndexErrors.DatabaseErrors.SqlNonTransientError.Reject(ex).asGrpcError
+
+  // PostgreSQL exceptions are handled based on the SQLState
+  // https://www.postgresql.org/docs/11/errcodes-appendix.html
+  private def isRetryablePsqlException(exception: PSQLException): Boolean =
+    exception.getSQLState match {
+      // Class 08 — Connection Exception
+      case state if state.startsWith("08") => true
+      // Failure to serialize db accesses, happens due to contention
+      case "40001" => true
+      // Retry on read only transaction, which can occur on Azure
+      case "25006" => true
+      // Retry on operator intervention errors, but not on `query_canceled` and `database_dropped`
+      case state if state.startsWith("57P") && state != "57014" && state != "57P04" => true
+      case _ => false
+    }
+
+  // Oracle exceptions are handled based on the error codes
+  // https://docs.oracle.com/en/database/oracle/oracle-database/19/errmg/ORA-00000.html#GUID-27437B7F-F0C3-4F1F-9C6E-6780706FB0F6
+  private def isRetryableOracleException(ex: SQLException): Boolean =
+    ex.getErrorCode match {
```

Comment:
Are the values returned from `getErrorCode` also visible as part of `getMessage`? Asking because I am asking myself if this valuable information will be available to us when troubleshooting.

Reply:
I would expect so.. But the ojdbc driver doesn't seem to be public and I have to navigate through decompiled code to understand.

Reply:
Hard to confirm from the decompiled code. But I remember seeing error codes in stack traces. Furthermore, the messages are unique and should be able to trace back to the codes (in case they are missing).

Reply:
:+1: